File: src/main/java/com/github/svegon/capi/event/render/LightmapDarknessFactorCallback.java
package com.github.svegon.capi.event.render;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;

public interface LightmapDarknessFactorCallback {
    Event<LightmapDarknessFactorCallback> EVENT = EventFactory.createArrayBacked(LightmapDarknessFactorCallback.class,
            (tickDelta, callback) -> {}, listeners -> (tickDelta, callback) -> {
        for (LightmapDarknessFactorCallback listener : listeners) {
            listener.getDarknessFactor(tickDelta, callback);

            if (callback.isCancelled()) {
                return;
            }
        }
            });

    void getDarknessFactor(float tickDelta, CallbackInfoReturnable<Float> callback);
}

File: src/main/kotlin/io/github/svegon/mclientapi/event/render/LightmapDarknessFactorCallback.kt
package io.github.svegon.capi.event.render

import io.github.svegon.capi.event.render.LightmapDarknessFactorCallback
import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable
import java.util.function.Function
import kotlin.Array
import kotlin.Float
import kotlin.invoke

interface LightmapDarknessFactorCallback {
    fun getDarknessFactor(tickDelta: Float, callback: CallbackInfoReturnable<Float?>?)

    companion object {
        @JvmField
        val EVENT: Event<LightmapDarknessFactorCallback?> = EventFactory.createArrayBacked(
            LightmapDarknessFactorCallback::class.java,
            LightmapDarknessFactorCallback { tickDelta: Float, callback: CallbackInfoReturnable<Float?>? -> },
            Function { listeners: Array<LightmapDarknessFactorCallback?>? ->
                LightmapDarknessFactorCallback { tickDelta: Float, callback: CallbackInfoReturnable<Float?>? ->
                    for (listener in listeners) {
                        listener.getDarknessFactor(tickDelta, callback)

                        if (callback.isCancelled()) {
                            return@LightmapDarknessFactorCallback
                        }
                    }
                }
            })
    }
}
